### PR TITLE
Bluetooth: Mesh: Enable Segmented Control Messages

### PR DESCRIPTION
**Pull request by AhmedHussein751**
Adding segmentation support to bt_mesh_ctl_send.

### CODE REVIEW

**Comment by joerchan** [comment]
@AhmedHussein751 Thank you :)

Whenever you fix changes in the commits in your PR you should amend the commits.
The checkpatch commit should be squashed together with the first commit.

Also you should use `git commit --signoff` (or just `-s`) to create the signoff line.

**Comment by AhmedHussein751** [comment]
Hi Joakim,

Thanks for the info. Should I soft reset to the commit and amend?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by joerchan** [comment]
@AhmedHussein751 You could use interactive rebase.
`git rebase -i HEAD~2`
Mark your first commit as `pick` and the second one as `squash`
You should end up with one commit then. This one you can append the signoff with `git commit --amend -s`

File: subsys/bluetooth/mesh/transport.c
Comment (WilliamGFish) [inline subsys/bluetooth/mesh/transport.c]: The segment count parameter is missing..Should it be 1?
tx doesn't have a seg_n (tx->seg_n)

File: subsys/bluetooth/mesh/transport.c
Comment (WilliamGFish) [inline subsys/bluetooth/mesh/transport.c]: variable declared twice

File: subsys/bluetooth/mesh/transport.c
Comment (WilliamGFish) [inline subsys/bluetooth/mesh/transport.c]: u8_t segment_count defined but never used

File: subsys/bluetooth/mesh/transport.c
Comment (WilliamGFish) [inline subsys/bluetooth/mesh/transport.c]: defined but never used

File: subsys/bluetooth/mesh/transport.c
Comment (WilliamGFish) [inline subsys/bluetooth/mesh/transport.c]: extra stray } which closes unopened block. Please remove.

File: subsys/bluetooth/mesh/transport.c
Comment (WilliamGFish) [inline subsys/bluetooth/mesh/transport.c]: Do you mean MIN rather than min?

**Review by WilliamGFish — CHANGES_REQUESTED** [review]
I haven't checked your logic but there are a number of errors that need resolving just a allow the code to compile.

**Comment by AhmedHussein751** [comment]
Sorry about that it seems there are some copy issues. I will fix them soon

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by joerchan** [comment]
@AhmedHussein751 You can always join the slack channel to get help with git if you need.

**Comment by carlescufi** [comment]
@jhedberg and @trond-snekvik can you please review this?

**Review by jhedberg — CHANGES_REQUESTED** [review]
@AhmedHussein751 sorry for the delay with reviewing this. Looks generally ok (based on a very quick look), but the deep indentation of `bt_mesh_ctl_send()` makes the code rather hard to read. I think it might make sense to add two helper functions that `bt_mesh_ctl_send()` calls, one for unsegmented and another for segmented messages, in order to reduce the function complexity and indentation levels.

One general comment on the coding style which would make it more consistent with the rest of the code base (or at least the Bluetooth subsystem): please add an empty line after every `}`, unless the next line is also a `}` (in which case no empty line in between), thanks.

**Review by trond-snekvik — APPROVED** [review]
This is the only permutation missing friend handling, but as there are no segmented control messages going to the LPN yet, I don't think this is a blocker

File: subsys/bluetooth/mesh/transport.c
Comment (jhedberg) [inline subsys/bluetooth/mesh/transport.c]: You can drop this empty line

File: subsys/bluetooth/mesh/transport.c
Comment (jhedberg) [inline subsys/bluetooth/mesh/transport.c]: Minor (and I think you just copied the same from the access message code): you could use `BIT_MASK(13)`

File: subsys/bluetooth/mesh/transport.c
Comment (jhedberg) [inline subsys/bluetooth/mesh/transport.c]: You can remove this empty line

File: subsys/bluetooth/mesh/transport.c
Comment (jhedberg) [inline subsys/bluetooth/mesh/transport.c]: The indentation of the split line is wrong here (which makes it look like the second line is a new statement rather than being a continuation of the parameter list to `bt_mesh_net_send()`

File: subsys/bluetooth/mesh/transport.c
Comment (jhedberg) [inline subsys/bluetooth/mesh/transport.c]: You can remove this empty line here. Generally: no empty line between an assignment to a variable and a test of the variable's value.

**Review by jhedberg — CHANGES_REQUESTED** [review]
Looks generally good, except for a couple more cosmetic issues. Once those are fixed I think we can merge this. Thanks!

**Review by jhedberg — APPROVED** [review]
@AhmedHussein751 I went and did some cleanups myself and did a push to your branch. Some things like indentation and line lengths are not that easy to see in the web interface. I also removed the inline declarations since it's better to leave that to the compiler (it will almost certainly inline any longer functions that are only called from a single place)

**Comment by AhmedHussein751** [comment]
> @AhmedHussein751 I went and did some cleanups myself and did a push to your branch. Some things like indentation and line lengths are not that easy to see in the web interface. I also removed the inline declarations since it's better to leave that to the compiler (it will almost certainly inline any longer functions that are only called from a single place)

Thanks, it is my first pull request to zephyr project and I had to fix a lot of issues in this commit. Great to get knowledge about the contribution process.

**Comment by jhedberg** [comment]
> Thanks, it is my first pull request to zephyr project and I had to fix a lot of issues in this commit. 
> Great to get knowledge about the contribution process.

No problem - I hope it won't be the last contribution from you ;)

Btw, a couple of questions:

1. Did you test that things still work for unsegmented messages (i.e. nothing broke there)?
2. What do you need this for? Some future spec version which introduces long control messages?

**Comment by AhmedHussein751** [comment]
> > Thanks, it is my first pull request to zephyr project and I had to fix a lot of issues in this commit.
> > Great to get knowledge about the contribution process.
> 
> No problem - I hope it won't be the last contribution from you ;)
> 
> Btw, a couple of questions:
> 
>     1. Did you test that things still work for unsegmented messages (i.e. nothing broke there)?
> 
>     2. What do you need this for? Some future spec version which introduces long control messages?

1. The first version merged was tested but unfortunately I don't have enough nodes to test the changes made during the pull request.

2. It was used before for mesh routing that violates the standard by using the restricted control messages opcodes. I decided to merge this part as it complies and maybe try to tweak the routing design in the future to fix the compliance issues.

File: subsys/bluetooth/mesh/transport.c
Comment (jhedberg) [inline subsys/bluetooth/mesh/transport.c]: I just realised this part is probably wrong. We don't do it for access layer messages either. IIRC even if the destination is a group address we still want the tx_seg timers and retransmissions to keep running, even though we won't receive any acks. @trond-snekvik can you confirm?

@AhmedHussein751 no need to push any changes regarding this - I'll do it myself (like I've done with the cleanups) when we reach consensus.

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/transport.c]: You're right, we still do periodic retransmissions for group messages

Comment (jhedberg) [reply subsys/bluetooth/mesh/transport.c]: Thanks for confirming. I've pushed an update now.

**Pull request merged**
